cpp-ch/local-engine/Functions/SparkFunctionArraysOverlap.cpp

```diff
@@ -0,0 +1,165 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <Columns/ColumnString.h>
+#include <Columns/ColumnNullable.h>
+#include <Functions/IFunction.h>
+#include <Functions/FunctionFactory.h>
+#include <DataTypes/DataTypeString.h>
+#include <DataTypes/DataTypesNumber.h>
+
+using namespace DB;
+
+namespace DB
+{
+namespace ErrorCodes
+{
+    extern const int ILLEGAL_TYPE_OF_ARGUMENT;
+    extern const int NUMBER_OF_ARGUMENTS_DOESNT_MATCH;
+}
+}
+
+namespace local_engine
+{
+class SparkFunctionArraysOverlap : public IFunction
+{
+public:
+    static constexpr auto name = "sparkArraysOverlap";
```

Comment:
can we contribute to upstream clickhouse first ?

Reply:
ok

Reply:
we could also contribute array_join to upstream. CH has strict code review and tests and it could help improve our code quality.

Reply:
CH has a similar function `hasAny`, but it different from `arraysOverlap`, has so many corner cases, it will be very complex to convert arraysOverlap, as below
```
parse arrays_overlap(arr1, arr2) as
                if (isNull(arr1) || isNull(arr2))
                    return NULL
                else if (isEmpty(arr1) || isEmpty(arr2))
                    return false;
                else if (arr1.hasAny(arr2))
                {
                    if (!arr1.has(NULL) || !arr2.has(NULL))
                        return true;
                    else if (arr1.intersect(arr2) != NULL)
                        return true
                    else
                        return NULL;
                }
                else if (arr1.has(NULL) || arr2.has(NULL))
                    return NULL;
                else if (arr1.has(NaN) && arr2.has(NaN))
                    return true;
                else
                    return false;
```

and it is not easy to slove the case when array parameters contain `NaN`. So I think it is better to write a self-defined function. @taiyang-li

Reply:
What about array_join ?

Reply:
ch has a same named function `array_join`, but its function different from this. and I'll make a contribute to CH.